### ISSUE TITLE
MueLu_UnitTestsTpetra_MPI_ tests timing out on ATDM cuda 9.2 builds on waterman, ride, and white

### ISSUE THREAD

**Issue by fryeguy52** [trilinos/Trilinos#3491]
CC: @trilinos/muelu , @srajama1  (Trilinos Linear Solvers Product Lead), @bartlettroscoe 
 
## Next Action Status

PR #3498 merged on 9/25/2018 which reduces cost of the expensive BlockCrs unit tests.  Next: Watch for passing non-timing out tests on 9/26/2018 ...

## Description
As shown in [this query](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2018-09-04&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=4&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=Trilinos-atdm-&field2=testname&compare2=65&value2=MueLu_UnitTestsTpetra_MPI_&field3=status&compare3=62&value3=Passed&field4=buildstarttime&compare4=83&value4=2018-09-20) the tests:
* MueLu_UnitTestsTpetra_MPI_1
* MueLu_UnitTestsTpetra_MPI_4

are failing often in the builds:
* Trilinos-atdm-white-ride-cuda-9.2-opt
* Trilinos-atdm-white-ride-cuda-9.2-debug

the test:
* MueLu_UnitTestsTpetra_MPI_4

is also failing every night on waterman in the builds:
* Trilinos-atdm-waterman-cuda-9.2-opt
* Trilinos-atdm-waterman-cuda-9.2-debug

All of the failures are from timeouts

## Steps to Reproduce on white

One should be able to reproduce this failure on the machine white as described in:
* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md
More specifically, the commands given for the system white are provided at:
* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#ridewhite
The exact commands to reproduce this issue should be:
```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh cuda-9.2-debug

$ cmake \
  -GNinja \
  -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
  -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_MueLu=ON \
  $TRILINOS_DIR

$ make NP=16

$ bsub -x -Is -q rhel7F -n 16 ctest -j16
```
## Steps to Reproduce on waterman

One should be able to reproduce this failure on the machine waterman as described in:
* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md
More specifically, the commands given for the system white are provided at:
* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#waterman
The exact commands to reproduce this issue should be:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh cuda-debug

$ cmake \
  -GNinja \
  -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
  -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_MueLu=ON \
  $TRILINOS_DIR

$ make NP=20

$ bsub -x -Is -n 20 ctest -j20
```

**Comment by jhux2** [comment]
On waterman, `MueLu_UnitTestsTpetra_MPI_4` takes 558 seconds.  This is actually 552 separate unit tests.  13 of the tests account for 338 seconds (70% of the total time):

```
91.7   Hierarchy_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_BlockCrs_UnitTest
57.3   IntrepidPCoarsenFactory_double_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_GenerateRepresentativeBasisNodes_HEX_Equispaced_UnitTest
53     IntrepidPCoarsenFactory_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_GenerateRepresentativeBasisNodes_HEX_Spectral_UnitTest
45.5   IntrepidPCoarsenFactory_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_GenerateRepresentativeBasisNodes_HEX_Equispaced_UnitTest
42.5   Hierarchy_double_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_BlockCrs_UnitTest
18.1   BlockedSmoother_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_Thyra_Nested_BGS_Setup_Apply_UnitTest
17.6   BlockedSmoother_double_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_Thyra_Nested_BGS_Setup_Apply_UnitTest
12.5   IntrepidPCoarsenFactory_double_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_GenerateRepresentativeBasisNodes_QUAD_Spectral_UnitTest
11     IntrepidPCoarsenFactory_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_GenerateRepresentativeBasisNodes_QUAD_Spectral_UnitTest
10.4   BlockedSmoother_double_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_NestedII01I2I_Thyra_BS_Setup_Apply2_UnitTest
10.1   IntrepidPCoarsenFactory_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_GenerateRepresentativeBasisNodes_QUAD_Equispaced_UnitTest
 9.57  IntrepidPCoarsenFactory_double_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_GenerateRepresentativeBasisNodes_QUAD_Equispaced_UnitTest
 9.21  BlockedSmoother_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_NestedII01I2I_Thyra_BS_Setup_Apply2_UnitTest
```


Question:  @fryeguy52  What is the current timeout?  My recollection is that the timeout is for the entire test run -- I don't know the dashboard vocab to be more specific.  Is this correct?

**Comment by bartlettroscoe** [comment]
> What is the current timeout? 

@jhux2, the timeout is 10 minutes per ctest.  You can split this up into multiple invocations if you want to.  I give the same justification for this approach that I gave in https://github.com/trilinos/Trilinos/issues/3417#issuecomment-420303404 which is:

The current test timeout is 10 minutes. Breaking a longer test into multiple tests is just fine. That was done to resolve #3256 for example.

This might seem like a silly thing to do but it has a good technical justification. No matter how may cores you throw at testing, you can never make the wall-clock time for testing take any less than your longest-running test. For example, if we increased the timeout of this test to 30 minutes and it takes 25 minutes to pass on average, then we can never finish testing in less than 25 minutes, no matter how many cores we throw at it. And many cores will sit idle while we wait for these few long-running (single MPI-process) tests to finish. Therefore, if we keep the timeout at 10 minutes, then we can always get testing done in about 10 minutes if throw enough cores at it.

If the tests for any individual package start taking a much larger percentage of the total processor time, then we can try to make things cheaper. You can see processor time in the "Proc Time" column in "Advanced View", for example, here which shows which packages are the processor hogs. If you sort by "Proc Time" it may surprise you what packages are at the top.

Please let me know if you have any questions about this.

But before we split up this set of unit tests into multiple ctest tests, can we look into why these tests take so much longer than on other CUDA builds on other platforms?  For example, take a look at [this query](https://testing.sandia.gov/cdash-dev-view/testSummary.php?project=1&name=MueLu_UnitTestsTpetra_MPI_4&date=2018-09-22) which shows all of the builds that run the test `MueLu_UnitTestsTpetra_MPI_4`.  And not that it is timing out in the 'waterman' build `Trilinos-atdm-waterman-cuda-9.2-opt` which is a fully optimized build.   The same build on 'white' takes 390s (as shown in that query).

**Comment by jhux2** [comment]
@bartlettroscoe Is this a duplicate of #3341?

**Comment by mhoemmen** [comment]
FYI it doesn't make sense to parallelize GPU tests unless they are sufficiently small _and_ you're running the MPS server on the GPU.

**Comment by bartlettroscoe** [comment]
> FYI it doesn't make sense to run GPU tests in parallel unless they are sufficiently small and you're running the MPS server on the GPU.

No, we see some speedup in the **overall** time to run the test if we run several tests at the same time.  I think we use `ctest -j8`.  I documented this in detail in one of the past issues or PRs.  We can find it if there is still a question about this.

**Comment by jhux2** [comment]
My patch definitely helped -- 687 tests now run, versus 558 previously (see [here](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2018-09-04&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=4&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=waterman&field2=testname&compare2=65&value2=MueLu_UnitTestsTpetra_MPI_&field3=status&compare3=62&value3=Passed&field4=buildstarttime&compare4=83&value4=2018-09-25)).  However, it's still timing out.  Looks like I need to do something about MueLu's Intrepid2 tests.

**Comment by bartlettroscoe** [comment]
Okay, looking at the `MueLu_UnitTest*` tests today in [this query](https://testing.sandia.gov/cdash-dev-view/queryTests.php?project=Trilinos&date=2018-09-27&filtercount=2&showfilters=1&filtercombine=or&field1=block&field1count=2&field1field1=buildname&field1compare1=65&field1value1=Trilinos-atdm-waterman-&field1field2=testname&field1compare2=65&field1value2=MueLu_UnitTests&field2=block&field2count=3&field2field1=buildname&field2compare1=65&field2value1=Trilinos-atdm-white-ride-&field2field2=site&field2compare2=61&field2value2=white&field2field3=testname&field2compare3=65&field2value3=MueLu_UnitTests) (using the new CDash OR filter feature), it looks like they all pass in all 4 waterman builds and all 5 builds on 'white' and ride and there are no failing MueLu tests at all in any of the promoted "ATDM" builds or in the 'waterman' builds as shown in [this query](https://testing.sandia.gov/cdash-dev-view/queryTests.php?project=Trilinos&date=2018-09-27&filtercount=2&showfilters=1&filtercombine=or&field1=block&field1count=3&field1field1=buildname&field1compare1=65&field1value1=Trilinos-atdm-waterman-&field1field2=testname&field1compare2=65&field1value2=MueLu_&field1field3=status&field1compare3=62&field1value3=passed&field2=block&field2count=3&field2field1=groupname&field2compare1=61&field2value1=ATDM&field2field2=testname&field2compare2=65&field2value2=MueLu_&field2field3=status&field2compare3=62&field2value3=passed).

@jhux2 thanks for taking care of this!

I think it is safe to close this issue now.
